pkg/models/job_history.go

```diff
@@ -49,5 +49,55 @@ type JobHistory struct {
 	ExecutionState *StateChange[ExecutionStateType] `json:"ExecutionState,omitempty"`
 	NewRevision    uint64                           `json:"NewRevision"`
 	Comment        string                           `json:"Comment,omitempty"`
+	Event          Event                            `json:"Event,omitempty"`
 	Time           time.Time                        `json:"Time"`
 }
+
+// Occurred returns when the action that triggered an update to job history
+// actually occurred.
+//
+// The Time field represents the moment that the JobHistory item was recorded,
+// i.e. it is almost always set to time.Now() when creating the object. This is
+// different to the Event.Timestamp which represents when the source of the
+// history update actually occurred.
+func (jh JobHistory) Occurred() time.Time {
+	if !jh.Event.Timestamp.Equal(time.Time{}) {
+		return jh.Event.Timestamp
+	}
+	return jh.Time
+}
+
+// EventTopic is a high level categorisation that can be applied to an event. It
+// should be a human-readable string with no dynamic content. They are used to
+// disambiguate events from the same component occurring in different contexts.
+// For example, an event emitted by S3 storage used as an input source and the
+// same event emitted by S3 storage used as a publisher would be tagged with
+// different topics.
+//
+// EventTopics do not need to conform to a centralized list – each module should
+// use event topics that make sense for their own logic. Event topics SHOULD be
+// unique.
+type EventTopic string
+
+// Event represents a progress report made by the system in its attempt to run a
+// job. Events are generated by the orchestrator and also passed back to the
+// orchestrator from the compute node.
+//
+// Events may be delivered in an async fashion – i.e, they may arrive much later
+// than the moment they occurred.
+type Event struct {
```

Comment:
Not sure I'd expect to find this in job_history.go as opposed to event.go - even started to add a comment that I couldn't find it.

Reply:
Yeah fair enough. I will move it.